File: rust/src/issue_comment.rs
//! Issue comment management.

use chrono::Utc;
use std::path::Path;
use uuid::Uuid;

use crate::agent_metadata::assign_agent_metadata_if_incomplete;
use crate::error::KanbusError;
use crate::event_history::{
    comment_payload, comment_updated_payload, now_timestamp, EventRecord, EventType,
};
use crate::issue_files::write_issue_to_file;
use crate::issue_lookup::load_issue_from_project;
use crate::issue_mutation::{persist_issue_mutation, PersistIssueMutationRequest};
use crate::models::{AgentMetadata, IssueComment, IssueData};
use crate::users::get_current_user;

/// Result of adding a comment to an issue.
#[derive(Debug, Clone)]
pub struct IssueCommentResult {
    pub issue: IssueData,
    pub comment: IssueComment,
}

fn generate_comment_id() -> String {
    Uuid::new_v4().to_string()
}

pub fn ensure_comment_ids(issue: &IssueData) -> (IssueData, bool) {
    let mut changed = false;
    let comments = issue
        .comments
        .iter()
        .map(|comment| {
            if comment.id.as_deref().unwrap_or("").is_empty() {
                changed = true;
                IssueComment {
                    id: Some(generate_comment_id()),
                    author: comment.author.clone(),
                    text: comment.text.clone(),
                    created_at: comment.created_at,
                    comment_type: comment.comment_type.clone(),
                    data: comment.data.clone(),
                    agent: comment.agent.clone(),
                }
            } else {
                comment.clone()
            }
        })
        .collect::<Vec<_>>();
    if !changed {
        return (issue.clone(), false);
    }
    (
        IssueData {
            comments,
            ..issue.clone()
        },
        true,
    )
}

fn normalize_prefix(prefix: &str) -> Result<String, KanbusError> {
    let trimmed = prefix.trim();
    if trimmed.is_empty() {
        return Err(KanbusError::IssueOperation(
            "comment id is required".to_string(),
        ));
    }
    Ok(trimmed.to_ascii_lowercase())
}

fn find_comment_by_prefix(issue: &IssueData, prefix: &str) -> Result<usize, KanbusError> {
    let normalized = normalize_prefix(prefix)?;
    let mut matches = Vec::new();
    for (index, comment) in issue.comments.iter().enumerate() {
        let Some(id) = comment.id.as_deref() else {
            continue;
        };
        if id.to_ascii_lowercase().starts_with(&normalized) {
            matches.push(index);
        }
    }
    match matches.len() {
        0 => Err(KanbusError::IssueOperation("comment not found".to_string())),
        1 => Ok(matches[0]),
        _ => {
            let ids = matches
                .iter()
                .filter_map(|index| issue.comments.get(*index))
                .filter_map(|comment| comment.id.as_deref())
                .map(|id| id.chars().take(6).collect::<String>())
                .collect::<Vec<_>>()
                .join(", ");
            Err(KanbusError::IssueOperation(format!(
                "comment id prefix is ambiguous; matches: {ids}"
            )))
        }
    }
}

fn persist_comment_mutation(
    root: &Path,
    lookup: &crate::issue_lookup::IssueLookupResult,
    updated_issue: IssueData,
    before_issue: IssueData,
    event_type: EventType,
    payload: serde_json::Value,
) -> Result<IssueData, KanbusError> {
    let actor_id = get_current_user();
    let event = EventRecord::new(
        updated_issue.identifier.clone(),
        event_type,
        actor_id.clone(),
        payload,
        now_timestamp(),
    );
    let event_id = event.event_id.clone();
    let result = persist_issue_mutation(&PersistIssueMutationRequest {
        project_dir: lookup.project_dir.clone(),
        issue_path: lookup.issue_path.clone(),
        issue: updated_issue,
        actor_id,
        events: vec![event],
        root: root.to_path_buf(),
        before_issue: Some(before_issue),
        relocate_to: None,
        regenerate_right_now: true,
    })?;
    if lookup.issue_path.parent() == Some(lookup.project_dir.join("issues").as_path()) {
        crate::gossip::publish_issue_mutation(
            root,
            &lookup.project_dir,
            &result.issue,
            Some(event_id),
            "issue.mutated",
        );
    }
    Ok(result.issue)
}

/// Add a comment to an issue.
///
/// # Arguments
/// * `root` - Repository root path.
/// * `identifier` - Issue identifier.
/// * `author` - Comment author.
/// * `text` - Comment text.
///
/// # Errors
/// Returns `KanbusError` if the issue cannot be found or updated.
pub fn add_comment(
    root: &Path,
    identifier: &str,
    author: &str,
    text: &str,
    agent: Option<crate::models::AgentMetadata>,
) -> Result<IssueCommentResult, KanbusError> {
    let lookup = load_issue_from_project(root, identifier)?;
    let timestamp = Utc::now();
    let comment = IssueComment {
        id: Some(generate_comment_id()),
        author: author.to_string(),
        text: Some(text.to_string()),
        created_at: timestamp,
        comment_type: "default".to_string(),
        data: std::collections::BTreeMap::new(),
        agent,
    };
    let (base_issue, _) = ensure_comment_ids(&lookup.issue);
    let mut comments = base_issue.comments.clone();
    comments.push(comment.clone());
    let updated = IssueData {
        comments,
        ..base_issue
    };
    let comment_id = comment
        .id
        .clone()
        .ok_or_else(|| KanbusError::IssueOperation("comment id is required".to_string()))?;
    let persisted = persist_comment_mutation(
        root,
        &lookup,
        updated,
        lookup.issue.clone(),
        EventType::CommentAdded,
        comment_payload(&comment_id, &comment.author, comment.agent.as_ref()),
    )?;
    Ok(IssueCommentResult {
        issue: persisted,
        comment,
    })
}

/// Ensure comment IDs exist for an issue and persist any changes.
pub fn ensure_issue_comment_ids(root: &Path, identifier: &str) -> Result<IssueData, KanbusError> {
    let lookup = load_issue_from_project(root, identifier)?;
    let (updated, changed) = ensure_comment_ids(&lookup.issue);
    if changed {
        write_issue_to_file(&updated, &lookup.issue_path)?;
    }
    Ok(updated)
}

/// Update an existing comment by id prefix.
pub fn update_comment(
    root: &Path,
    identifier: &str,
    comment_id_prefix: &str,
    text: Option<&str>,
    agent: Option<AgentMetadata>,
) -> Result<IssueData, KanbusError> {
    let lookup = load_issue_from_project(root, identifier)?;
    let (mut issue, _) = ensure_comment_ids(&lookup.issue);
    let index = find_comment_by_prefix(&issue, comment_id_prefix)?;
    let existing_comment = issue
        .comments
        .get(index)
        .cloned()
        .ok_or_else(|| KanbusError::IssueOperation("comment not found".to_string()))?;
    let assigned_agent =
        assign_agent_metadata_if_incomplete(existing_comment.agent.clone(), agent)?;
    let mut changed = false;
    if let Some(comment) = issue.comments.get_mut(index) {
        if let Some(new_text) = text {
            comment.text = Some(new_text.to_string());
            changed = true;
        }
        if assigned_agent != comment.agent {
            comment.agent = assigned_agent;
            changed = true;
        }
    }
    if !changed {
        return Err(KanbusError::IssueOperation(
            "no comment updates requested".to_string(),
        ));
    }
    let comment_id = existing_comment
        .id
        .clone()
        .ok_or_else(|| KanbusError::IssueOperation("comment id is required".to_string()))?;
    persist_comment_mutation(
        root,
        &lookup,
        issue,
        lookup.issue.clone(),
        EventType::CommentUpdated,
        comment_updated_payload(&comment_id, &existing_comment.author),
    )
}

/// Delete an existing comment by id prefix.
pub fn delete_comment(
    root: &Path,
    identifier: &str,
    comment_id_prefix: &str,
) -> Result<IssueData, KanbusError> {
    let lookup = load_issue_from_project(root, identifier)?;
    let (mut issue, _) = ensure_comment_ids(&lookup.issue);
    let index = find_comment_by_prefix(&issue, comment_id_prefix)?;
    let removed = issue.comments.remove(index);
    let comment_id = removed
        .id
        .clone()
        .ok_or_else(|| KanbusError::IssueOperation("comment id is required".to_string()))?;
    persist_comment_mutation(
        root,
        &lookup,
        issue,
        lookup.issue.clone(),
        EventType::CommentDeleted,
        comment_payload(&comment_id, &removed.author, removed.agent.as_ref()),
    )
}

File: python/src/kanbus/issue_comment.py
"""Issue comment management."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from uuid import uuid4

from kanbus.agent_metadata import (
    AgentMetadataResolutionError,
    assign_agent_metadata_if_incomplete,
)

from kanbus.event_history import (
    comment_payload,
    comment_updated_payload,
    create_event,
    now_timestamp,
)
from kanbus.gossip import publish_issue_mutation
from kanbus.issue_files import write_issue_to_file
from kanbus.issue_lookup import IssueLookupError, load_issue_from_project
from kanbus.issue_mutation import PersistIssueMutationRequest, persist_issue_mutation
from kanbus.models import AgentMetadata, IssueComment, IssueData
from kanbus.users import get_current_user


class IssueCommentError(RuntimeError):
    """Raised when issue comment creation fails."""


@dataclass(frozen=True)
class IssueCommentResult:
    """Result of adding a comment to an issue."""

    issue: IssueData
    comment: IssueComment


def _generate_comment_id() -> str:
    return str(uuid4())


def _ensure_comment_ids(issue: IssueData) -> tuple[IssueData, bool]:
    changed = False
    comments = []
    for comment in issue.comments:
        if not comment.id:
            changed = True
            comments.append(
                IssueComment(
                    id=_generate_comment_id(),
                    author=comment.author,
                    text=comment.text,
                    created_at=comment.created_at,
                    comment_type=comment.comment_type,
                    data=comment.data,
                    agent=comment.agent,
                )
            )
        else:
            comments.append(comment)
    if not changed:
        return issue, False
    updated = issue.model_copy(update={"comments": comments})
    return updated, True


def _normalize_prefix(prefix: str) -> str:
    normalized = prefix.strip().lower()
    if not normalized:
        raise IssueCommentError("comment id is required")
    return normalized


def _find_comment_index(issue: IssueData, prefix: str) -> int:
    normalized = _normalize_prefix(prefix)
    matches: list[int] = []
    for index, comment in enumerate(issue.comments):
        if comment.id and comment.id.lower().startswith(normalized):
            matches.append(index)
    if not matches:
        raise IssueCommentError("comment not found")
    if len(matches) > 1:
        ids = ", ".join(
            (issue.comments[index].id or "")[:6]
            for index in matches
            if issue.comments[index].id
        )
        raise IssueCommentError(f"comment id prefix is ambiguous; matches: {ids}")
    return matches[0]


def _persist_comment_mutation(
    root: Path,
    lookup,
    updated_issue: IssueData,
    before_issue: IssueData,
    event_type: str,
    payload: dict,
) -> IssueData:
    actor_id = get_current_user()
    event = create_event(
        issue_id=updated_issue.identifier,
        event_type=event_type,
        actor_id=actor_id,
        payload=payload,
        occurred_at=now_timestamp(),
    )
    try:
        result = persist_issue_mutation(
            PersistIssueMutationRequest(
                project_dir=lookup.project_dir,
                issue_path=lookup.issue_path,
                issue=updated_issue,
                actor_id=actor_id,
                events=[event],
                before_issue=before_issue,
                root=root,
            )
        )
    except Exception as error:  # noqa: BLE001
        raise IssueCommentError(str(error)) from error
    if lookup.issue_path.parent == lookup.project_dir / "issues":
        publish_issue_mutation(
            root,
            lookup.project_dir,
            result.issue,
            event.event_id,
            "issue.mutated",
        )
    return result.issue


def add_comment(
    root: Path,
    identifier: str,
    author: str,
    text: str,
    agent: Optional[AgentMetadata] = None,
) -> IssueCommentResult:
    """Add a comment to an issue.

    :param root: Repository root path.
    :type root: Path
    :param identifier: Issue identifier.
    :type identifier: str
    :param author: Comment author.
    :type author: str
    :param text: Comment text.
    :type text: str
    :param agent: Optional agent provenance metadata for this comment.
    :type agent: Optional[AgentMetadata]
    :return: Comment result including the updated issue.
    :rtype: IssueCommentResult
    :raises IssueCommentError: If the issue cannot be found or updated.
    """
    try:
        lookup = load_issue_from_project(root, identifier)
    except IssueLookupError as error:
        raise IssueCommentError(str(error)) from error

    timestamp = datetime.now(timezone.utc)
    base_issue, _ = _ensure_comment_ids(lookup.issue)
    comment = IssueComment(
        id=_generate_comment_id(),
        author=author,
        text=text,
        created_at=timestamp,
        agent=agent,
    )
    comments = [*base_issue.comments, comment]
    updated = base_issue.model_copy(update={"comments": comments})
    comment_id = comment.id
    if not comment_id:
        raise IssueCommentError("comment id is required")
    persisted = _persist_comment_mutation(
        root,
        lookup,
        updated,
        lookup.issue,
        "comment_added",
        comment_payload(comment_id, comment.author, comment.agent),
    )
    return IssueCommentResult(issue=persisted, comment=comment)


def ensure_issue_comment_ids(root: Path, identifier: str) -> IssueData:
    """Ensure comment ids are set for an issue and persist any changes."""
    try:
        lookup = load_issue_from_project(root, identifier)
    except IssueLookupError as error:
        raise IssueCommentError(str(error)) from error
    updated, changed = _ensure_comment_ids(lookup.issue)
    if changed:
        write_issue_to_file(updated, lookup.issue_path)
    return updated


def update_comment(
    root: Path,
    identifier: str,
    comment_id: str,
    text: Optional[str] = None,
    agent: Optional[AgentMetadata] = None,
) -> IssueData:
    """Update a comment by id prefix.

    :param root: Repository root path.
    :type root: Path
    :param identifier: Issue identifier.
    :type identifier: str
    :param comment_id: Comment id or prefix.
    :type comment_id: str
    :param text: Replacement comment text, or None to leave text unchanged.
    :type text: Optional[str]
    :param agent: Agent provenance to set when the comment is incomplete.
    :type agent: Optional[AgentMetadata]
    :return: Persisted issue.
    :rtype: IssueData
    :raises IssueCommentError: If the update fails.
    """
    try:
        lookup = load_issue_from_project(root, identifier)
    except IssueLookupError as error:
        raise IssueCommentError(str(error)) from error
    issue, _ = _ensure_comment_ids(lookup.issue)
    index = _find_comment_index(issue, comment_id)
    comments = list(issue.comments)
    existing_comment = comments[index]
    try:
        assigned_agent = assign_agent_metadata_if_incomplete(
            existing_comment.agent, agent
        )
    except AgentMetadataResolutionError as error:
        raise IssueCommentError(str(error)) from error
    update_fields = {}
    if text is not None:
        update_fields["text"] = text
    if assigned_agent != existing_comment.agent:
        update_fields["agent"] = assigned_agent
    if not update_fields:
        raise IssueCommentError("no comment updates requested")
    updated_comment = comments[index].model_copy(update=update_fields)
    comments[index] = updated_comment
    updated = issue.model_copy(update={"comments": comments})
    if not existing_comment.id:
        raise IssueCommentError("comment id is required")
    return _persist_comment_mutation(
        root,
        lookup,
        updated,
        lookup.issue,
        "comment_updated",
        comment_updated_payload(existing_comment.id, existing_comment.author),
    )


def delete_comment(root: Path, identifier: str, comment_id: str) -> IssueData:
    """Delete a comment by id prefix."""
    try:
        lookup = load_issue_from_project(root, identifier)
    except IssueLookupError as error:
        raise IssueCommentError(str(error)) from error
    issue, _ = _ensure_comment_ids(lookup.issue)
    index = _find_comment_index(issue, comment_id)
    comments = list(issue.comments)
    removed_comment = comments.pop(index)
    updated = issue.model_copy(update={"comments": comments})
    if not removed_comment.id:
        raise IssueCommentError("comment id is required")
    return _persist_comment_mutation(
        root,
        lookup,
        updated,
        lookup.issue,
        "comment_deleted",
        comment_payload(
            removed_comment.id, removed_comment.author, removed_comment.agent
        ),
    )
